### DISCUSSION TITLE
Lint rule idea noDynamicNamespaceImportAccess

### DISCUSSION THREAD

**Discussion by minht11** [biomejs/biome#3164] [Rule suggestion]
I haven't found any rule like that when I searched for it briefly, so this might be new rule.
Disallow dynamic access of namespaced imports. For example we use [ParaglideJS](https://github.com/opral/inlang-paraglide-js) and it export its i18n keys like this:
```ts
import * as m from './messages'
```
Under the hood each key is separate file, so Rollup is able to tree shake them, however if we use it as dynamic map:
```ts
m[dynamicKey]
```
The whole object must be inlined and bundle size would explode.
In this more contrived example, people might pull whole react lib instead of one hook:
```ts
import * as React from 'react'

React['useState']
```

So I would propose adding new lint rule, which disallows dynamic access on namespaced imports, this would be less restrictive version of https://biomejs.dev/linter/rules/no-namespace-import/

**Comment by Conaclos** [comment]
[useLiteralKeys](https://biomejs.dev/linter/rules/use-literal-keys) will report this pattern. Do you think we really need a specific rule here? If we add this rule, I think we should still allow computed key that are not identifiers such as `React["@special"]`.

**Reply by minht11** [reply]
Will it report cases like this:
```ts
import * as m from './messages'

const someComputedKeyOrEvenExternalProperty = ...

const value = m[someComputedKeyOrEvenExternalProperty]
```

Its not only about literals, its about any non static property access which bundler would not be able to treeshake.

**Reply by Conaclos** [reply]
You are right. Look like a good rule. Feel free to open an issue to implement this.

**Reply by minht11** [reply]
I just remembered that we use [auto import](https://github.com/unplugin/unplugin-auto-import), so for my case it would not really help, because code like this:
```ts
import * as m from './messages'
const value =  m[someComputedKeyOrEvenExternalProperty]
```
becomes:
```ts
const value =  m[someComputedKeyOrEvenExternalProperty]
```
variable `m` is auto imported. So my proposed rule would not work.

While I think rule by itself is useful, what do you think adding general rule `noDynamicPropertyAccess` which would allow user to specify custom object names, which should disallow such access, it also could have option to disallow namespaced import access.

1. Do you think `noDynamicPropertyAccess` rule could be added to biome? With option to customize what things to match?
2. If general rule is added should `noDynamicNamespaceImportAccess` be merged under it or remain separate?
